[WIP] make the type of extract richer

divarvel opened the pull request:
See #387 for the whole context. The main idea is to make `extract` return a more detailed errors (currently, it returns a `Maybe`).

Following discussion with @ocharles, and based on @sellout's comments in the issue, I went with a `Validation (NonEmpty ExtractError)`, where `ExtractError` is either `InvalidType` or a freeform text for user-supplied errors. Reading the tests, the example provided by @sellout seems at odds with the contact that `extract` should only fail on type errors.

For now it does not track line number or anything else, but thanks to the applicative instance, it accumulates errors for lists and records. Instead of providing a monad instance through a newtype, I added a `validate` function which is essentially `>>=`.

I've added dependencies to `either` (for `Validation`) and to `semigroupoids` (for `Data.Functor.Alt`, as `NonEmptyList a` only forms a semigroup).

# ToDo

- [x] add a proper error message for the errors that are not type mismatches
- [x] make tests pass (there is an error message mismatch)

Gabriel439 commented on dhall/src/Dhall.hs:
```diff
@@ -96,13 +99,16 @@ module Dhall
 import Control.Monad (guard)
 import Data.Coerce (coerce)
+import Data.Either.Validation (Validation(..))
+import Data.Functor.Alt ((<!>))
```
I believe you can use `<|>` instead of `<!>`, that way you don't need to directly depend on `semigroupoids`

divarvel replied:
I've replaced `<|>` with `<!>` because it requires a `Monoid`, whereas `NonEmpty a` (and `Validation (NonEmpty a) b`) only have a `Semigroup`. If the dependency on semigroupoids is a problem, I can just reimplement `<!>` instead of pulling the library.

Gabriel439 replied:
Oh, alright.  Then you can leave the `(<!>)`.  Minimizing the constraint is a good reason to use it

Gabriel439 commented on dhall/src/Dhall.hs:
```diff
@@ -157,12 +165,47 @@ import qualified Dhall.Util
+  } deriving Semigroup
+
+typeError :: Expr s a -> Expr s a -> Extractor s a b
+typeError expected actual = Failure . ExtractErrors . (:| []) . TypeMismatch $ InvalidType expected actual
```
```suggestion
typeError expected actual = Failure . ExtractErrors . pure . TypeMismatch $ InvalidType expected actual
```

Gabriel439 commented on dhall/src/Dhall.hs:
```diff
@@ -157,12 +165,47 @@ import qualified Dhall.Util
+typeError expected actual = Failure . ExtractErrors . (:| []) . TypeMismatch $ InvalidType expected actual
+
+extractError :: Text -> Extractor s a b
+extractError = Failure . ExtractErrors . (:| []) . ExtractError
```
```suggestion
extractError = Failure . ExtractErrors . pure . ExtractError
```

Gabriel439 commented on dhall/src/Dhall.hs:
```diff
@@ -157,12 +165,47 @@ import qualified Dhall.Util
+
+-- | opt-in monadic composition for `Extractor`. This is done this way to
+-- keep applicative behaviour on @ExtractErrors@
+validate :: Extractor s a b -> (b -> Extractor s a c) -> Extractor s a c
```
If we don't use `validate` internally I would omit it from the API, especially given the fact that we expose the `Extractor` type's internals

Gabriel439 replied:
I think it's fine if people use this for non-type-errors.

I think my main issue with this function is that it's basically a `Monad` instance but without the instance.  I think if we firmly believe that the user will need `Monad` operations then we should go all the way and provide a `newtype` that exposes a `Monad` instance rather than a named bind function, because the latter will not work with `do` notation, monad transformers, or `Control.Monad` utilities.

divarvel replied:
Ok, I'll make a monad instance with a newtype then.

divarvel replied:
Now that I think of it, `Extractor s a` is just a type alias for `Validation (NonEmpty (ExtractError s a)) b`. I can add a new type alias for `Either (NonEmpty (ExtractError s a)) b` (named eg `MonadicExtractError`), and maybe provide a pair of  `from/to` functions (aliases of `eitherToValidation` and `validationToEither`). Or we just let implementors use `Data.Either.Validation` functions themselves

Gabriel439 commented on dhall/src/Dhall.hs:
```diff
@@ -157,12 +165,47 @@ import qualified Dhall.Util
+validate (Failure es) _ = Failure es
+
+instance (Pretty s, Pretty a, Typeable s, Typeable a) => Show (ExtractErrors s a) where
+    show (ExtractErrors es)  = unlines . Data.List.NonEmpty.toList . fmap show $ es
```
Two suggestions for formatting the error output:

* If there is more than one error, prefix all of the errors with a brief message saying that there multiple errors
* Separate the errors visually (i.e. perhaps add a blank line between them if there's not one already)

divarvel replied:
:+1: , i'll do that

Gabriel439 commented on dhall/src/Dhall.hs:
```diff
@@ -362,9 +405,9 @@ inputWithSettings settings (Type {..}) txt = do
-                     (InvalidType expected expr')
+        Success x  -> return x
+        Failure e -> Control.Exception.throwIO
+                     e
```
The `e` no longer needs to be on its own line

divarvel replied:
Indeed, that's fixed

divarvel commented:
As suggested, i've special-cased the error message for a single error (so that it doesn't change current behaviour). I've added a small message when there are multiple errors, and added an error message for extraction errors.

Once we've found a satisfying solution with `validate`, I can squash my commits.

divarvel commented:
I can drop semigroupoids and use `ealt` instead of `<!>`.

divarvel commented:
I've downgraded deps in stack-lts-6.yaml, but there are still issues with transformers-compat.
I've tried removing semigroupoids as well, but it still fails. I'm not sure what I can try next.

Gabriel439 commented:
@divarvel: I'll take a look at it tonight to see if I can get it building with `stack`'s LTS6 resolver

divarvel commented:
Thanks!

Gabriel439 commented:
Alright, I have a pull request up against your branch that I believe should fix the LTS 6 build failures: https://github.com/divarvel/dhall-haskell/pull/1

divarvel commented:
Thanks a lot! It looks like it's fixed on LTS-6. I've rebased on latest master, so unless something else is missing, I can shash the commits.

Gabriel439 commented:
@divarvel: You're welcome!  There's no need to rebase or squash ahead of time because the repository settings enforce that we "Squash and merge" anyway when merging pull requests
